patches/tModLoader/Terraria/ModLoader/DamageClassLoader.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Terraria.ModLoader
+{
+	public static class DamageClassLoader
+	{
+		public static int DamageClassCount => damageClasses.Count;
+
+		internal static List<ModDamageClass> damageClasses = new List<ModDamageClass>();
+
+		internal static bool DamageClassExists(ModDamageClass damageClass) => damageClass == null ? false : damageClasses.Contains(damageClass);
+
+		internal static void Add(ModDamageClass damageClass) {
+			damageClasses.Add(damageClass);
+		}
+
+		internal static int GetIndex<T>() => damageClasses.FindIndex(d => d.GetType() == typeof(T));
+
+		internal static int GetIndex(ModDamageClass damageClass) => damageClasses.FindIndex(d => d == damageClass);
```

Comment:
When I said that there's no need for a public id on damage classes, I meant just making it internal. Using linq could potentially slow everything down. Bring back the index integer in an internal form, and use it instead of this. You can replace places that use the generic method with ModContent.GetInstance<T>().index